src/transform/src/redundant_join.rs

```diff
@@ -25,95 +38,399 @@ impl crate::Transform for RedundantJoin {
+                            prov.binding
+                                .iter()
+                                .find(|(_src, inp)| inp == c)
+                                .map(|(src, _inp)| (*src, i))
+                        })
+                        .collect::<Vec<_>>();
+                    prov.binding = new_bindings;
+                }
+                result.retain(|p| !p.binding.is_empty());
+                // TODO: For min, max aggregates, we could preserve provenance
+                // if the expression references a column. We would need to un-set
+                // the `exact` bit in that case, and so we would want to keep both
+                // sets of provenance information.
+                result
             }
+
+            RelationExpr::Threshold { input } => {
+                // Threshold may drop records, and so we unset `exact`.
+                let mut result = self.action(input, lets);
+                for prov in result.iter_mut() {
+                    prov.exact = false;
+                }
+                result
+            }
+
+            RelationExpr::TopK { input, .. } => {
+                // TopK may drop records, and so we unset `exact`.
+                let mut result = self.action(input, lets);
+                for prov in result.iter_mut() {
+                    prov.exact = false;
+                }
+                result
+            }
+
+            RelationExpr::Project { input, outputs } => {
+                // Projections re-order, drop, and duplicate columns,
+                // but they neither drop rows nor invent values.
+                let mut result = self.action(input, lets);
+                for provenance in result.iter_mut() {
+                    let new_binding = (0..outputs.len())
+                        .enumerate()
+                        .flat_map(|(i, c)| {
+                            provenance
+                                .binding
+                                .iter()
+                                .find(|(_, l)| l == &c)
+                                .map(|(s, _)| (*s, i))
+                        })
+                        .collect::<Vec<_>>();
+
+                    provenance.binding = new_binding;
+                }
+                result
+            }
+
+            RelationExpr::FlatMap { input, .. } => {
+                // FlatMap may drop records, and so we unset `exact`.
+                let mut result = self.action(input, lets);
+                for prov in result.iter_mut() {
+                    prov.exact = false;
+                }
+                result
+            }
+
+            RelationExpr::Negate { input } => {
+                // Negate changes the sign on its multiplicities,
+                // which means "distinct" counts would now be -1.
+                // We set `exact` to false to inhibit the optimization,
+                // but should probably fix `.keys` instead.
```

Comment:
Is this hard to do?

Reply:
No, probably not. I mean, easy to break it, and I doubt we have anyone relying on the keys of `x.negate().negate()`. I polled Slack for takes on this and folks seemed to go in a different direction, discussion-wise.